# Projeto #1 - Mineração de padrões frequentes

<p style='text-align: justify;'><font size=3.>Análise de acidentes nas rodovias federais brasileira entre 2007 e 2021.</font></p>

> <p style='text-align: justify;'><font size=2.85>Esse projeto utiliza a versão 0.17.0 do mlxtend.</font></p>

In [2]:
import mlxtend
print(mlxtend.__version__)

0.17.0


### Bibliotecas básicas e outros imports

In [3]:
import warnings
import numpy as np
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import association_rules
from mlxtend.frequent_patterns import fpgrowth

warnings.filterwarnings('ignore')
pd.set_option('display.max_rows', 1000)
pd.set_option('display.max_columns', 200)
np.set_printoptions(suppress=True, precision=3)

%run modules/text.py

### Carregamento dos dados

In [4]:
df = pd.read_csv('dataset/cleaned_acidentes_rodovias_brasileiras_2007_a_2021.csv', low_memory=False, encoding='utf-8')
print(f'Linhas: {df.shape[0]} | Coluna:  {df.shape[1]}')
df.head(1)

Linhas: 1894736 | Coluna:  24


id dia_semana  uf   br     municipio causa_acidente  \
0  83123214.0     quarta  SC  280  Rio negrinho         outras   

    tipo_acidente classificacao_acidente     fase_dia  sentido_via  \
0  saida de pista               ignorada  plena noite  decrescente   

  condicao_metereologica tipo_pista tracado_via uso_solo tipo_veiculo  \
0                  chuva    simples       curva   urbano    automovel   

  tipo_envolvido estado_fisico        data  mes   ano estacao_do_ano  \
0       condutor      ignorada  2014-01-01    1  2014          verão   

  regiao_pais  estacao  clima  
0         sul        3      1

### Remoção valores inválidos

In [5]:
df.drop(df[df['tipo_veiculo'] == 'nao identificado'].index, inplace=True)
df.drop(df[df['tipo_veiculo'] == 'nao informado'].index, inplace=True)
df.drop(df[df['classificacao_acidente'] == 'ignorada'].index, inplace=True)
df.drop(df[df['condicao_metereologica'] == 'ignorada'].index, inplace=True)
print(f'Linhas: {df.shape[0]} | Coluna:  {df.shape[1]}')

Linhas: 1876750 | Coluna:  24


### Geração de features

#### Agrupamento tipos de acidentes

In [6]:
df['tipo_acidente'] = df['tipo_acidente'].apply(mapTiposDeAcidentes)

#### Agrupamento fase do dia

In [7]:
df['fase_dia'] = df['fase_dia'].map(mapFase)

#### Agrupamento dias da semana

In [8]:
df['dia_semana'] = df['dia_semana'].apply(mapDiasDaSemana)

#### Agrupamento condição meteorológica

In [9]:
df['condicao_clima'] = df['condicao_metereologica'].map(mapClima)

### Agrupamento tamanho veículos

In [10]:
df['porte_veiculo'] = df['tipo_veiculo'].apply(mapTamanhoVeiculos)

### Seleção dos recursos para análise

In [11]:
cols = [
        'classificacao_acidente',
        'tipo_acidente', 
        'fase_dia',	
        'condicao_clima', 
        'tracado_via',
        'porte_veiculo', 
        'estacao_do_ano',
        'dia_semana'
        ]
data = df[cols]
data.head()

classificacao_acidente  tipo_acidente fase_dia condicao_clima tracado_via  \
1            sem vitimas        colisão    noite     tempo ruim        reta   
2            sem vitimas  atropelamento    noite     ensolarado        reta   
3            sem vitimas        colisão    noite     ensolarado  cruzamento   
4    com vitimas feridas        colisão    noite        nublado        reta   
5            sem vitimas        colisão    noite        nublado        reta   

   porte_veiculo estacao_do_ano dia_semana  
1  pequeno porte          verão   dia útil  
2  pequeno porte          verão   dia útil  
3  pequeno porte          verão   dia útil  
4  pequeno porte          verão   dia útil  
5  pequeno porte          verão   dia útil

### Agrupamento

#### 1º cluster: Sem vítimas

In [12]:
sem_vitimas = data.query('classificacao_acidente=="sem vitimas"')
sem_vitimas.drop(['classificacao_acidente'], axis=1, inplace=True)
print(f'Linhas: {sem_vitimas.shape[0]} | Coluna:  {sem_vitimas.shape[1]}')

Linhas: 968151 | Coluna:  7


#### 2º cluster: Vítimas feridas



In [13]:
vitimas_feridas = data.query('classificacao_acidente=="com vitimas feridas"')
vitimas_feridas.drop(['classificacao_acidente'], axis=1, inplace=True)
print(f'Linhas: {vitimas_feridas.shape[0]} | Coluna:  {vitimas_feridas.shape[1]}')

Linhas: 797703 | Coluna:  7


#### 3º cluster: Vítimas fatais

In [14]:
vitimas_fatais = data.query('classificacao_acidente=="com vitimas fatais"')
vitimas_fatais.drop(['classificacao_acidente'], axis=1, inplace=True)
print(f'Linhas: {vitimas_fatais.shape[0]} | Coluna:  {vitimas_fatais.shape[1]}')

Linhas: 78625 | Coluna:  7


## Mineração das regras

#### Enconder da base

In [15]:
def transactionEncoderData(dados:np.ndarray) -> pd.DataFrame:
    transEncoder = TransactionEncoder()
    a_data = transEncoder.fit(dados).transform(dados)
    dados = pd.DataFrame(a_data, columns=transEncoder.columns_)
    return dados*1

#### Itens frequentes
> <p style='text-align: justify;'><font size=2.85>Fpgrowth foi escolhido por ser mais eficiente</font></p>

In [16]:
def getFrequentItemsets(encoderData:pd.DataFrame, minsup:float=0.3, useColnames:bool=True) -> pd.DataFrame:
    return fpgrowth(encoderData, min_support=minsup, use_colnames=useColnames)

#### Regras

In [17]:
def getAssociationRules(itemsets:pd.DataFrame, metric:str='lift', minThreshold:int=1) -> pd.DataFrame:
    return association_rules(itemsets, metric=metric, min_threshold=minThreshold)

### Acidentes sem vítimas

In [18]:
sem_vitimas = transactionEncoderData(sem_vitimas.values)
frequent_itemsets = getFrequentItemsets(sem_vitimas)
rules_sem_vitimas = getAssociationRules(frequent_itemsets)
rules_sem_vitimas = rules_sem_vitimas.sort_values(by='lift', ascending=False)
rules_sem_vitimas[['antecedents', 'consequents', 'confidence', 'lift', 'leverage', 'conviction']].query('confidence>=0.70').reset_index(drop=True)

antecedents          consequents  confidence  \
0                    (ensolarado, dia)  (dia útil, colisão)    0.720041   
1                          (dia, reta)  (dia útil, colisão)    0.718397   
2                (dia, dia útil, reta)            (colisão)    0.805858   
3          (ensolarado, dia útil, dia)            (colisão)    0.805672   
4      (ensolarado, dia útil, colisão)                (dia)    0.751962   
5                          (dia, reta)            (colisão)    0.798857   
6                    (ensolarado, dia)            (colisão)    0.797706   
7         (ensolarado, dia útil, reta)            (colisão)    0.792497   
8          (ensolarado, pequeno porte)            (colisão)    0.791467   
9                (ensolarado, colisão)                (dia)    0.737754   
10     (dia útil, pequeno porte, reta)            (colisão)    0.788603   
11               (ensolarado, colisão)               (reta)    0.785558   
12     (ensolarado, dia útil, colisão)               (reta)    0.785403   
13                  (ensolarado, reta)            (colisão)    0.785705   
14               (pequeno porte, reta)            (colisão)    0.781354   
15         (ensolarado, pequeno porte)               (reta)    0.769513   
16              (ensolarado, dia útil)            (colisão)    0.768169   
17              (ensolarado, dia útil)                (dia)    0.716959   
18                    (dia útil, reta)            (colisão)    0.766226   
19              (ensolarado, dia útil)               (reta)    0.761292   
20                        (ensolarado)               (reta)    0.760412   
21            (dia, dia útil, colisão)               (reta)    0.759702   
22                   (ensolarado, dia)               (reta)    0.759582   
23                 (dia útil, colisão)               (reta)    0.759441   
24                        (ensolarado)            (colisão)    0.760555   
25                              (reta)            (colisão)    0.760012   
26                           (colisão)               (reta)    0.758240   
27                      (dia, colisão)               (reta)    0.757710   
28           (dia útil, colisão, reta)                (dia)    0.709358   
29                 (dia útil, colisão)                (dia)    0.709115   
30                (dia, pequeno porte)            (colisão)    0.758031   
31  (dia útil, pequeno porte, colisão)               (reta)    0.752937   
32            (pequeno porte, colisão)               (reta)    0.751971   
33                        (ensolarado)                (dia)    0.703396   
34                     (dia, dia útil)            (colisão)    0.751670   
35                  (ensolarado, reta)                (dia)    0.702628   
36                               (dia)            (colisão)    0.742985   
37           (dia útil, pequeno porte)            (colisão)    0.739809   
38          (ensolarado, colisão, dia)           (dia útil)    0.902639   
39                (dia, colisão, reta)           (dia útil)    0.899281   
40                      (dia, colisão)           (dia útil)    0.896922   
41                     (pequeno porte)            (colisão)    0.731682   
42                   (ensolarado, dia)           (dia útil)    0.893714   
43                         (dia, reta)           (dia útil)    0.891468   
44                               (dia)           (dia útil)    0.886560   
45               (ensolarado, colisão)           (dia útil)    0.885585   
46         (ensolarado, colisão, reta)           (dia útil)    0.885410   
47                     (colisão, reta)           (dia útil)    0.880033   
48                          (dia útil)            (colisão)    0.718618   
49                           (colisão)           (dia útil)    0.878641   
50                  (ensolarado, reta)           (dia útil)    0.877822   
51                        (ensolarado)           (dia útil)    0.876808   
52                          (dia útil)              

> <p style='text-align: justify;'><font size=2.85>Nos resultados da análise das regras de associação para acidentes sem vítimas, a regra com o maior valor de Lift é 1.15 e mostra que há, principalmente, colisões que ocorrem em dias ensolarados durante a semana. Avaliando as demais regras, temos que a maioria dos acidentes ocorrem durante o dia e envolvem veículos de pequeno porte sendo o tipo de acidente mais comum as colisões em pistas retas.</font></p>

### Acidentes com vítimas feridas

In [19]:
vitimas_feridas = transactionEncoderData(vitimas_feridas.values)
frequent_itemsets = getFrequentItemsets(vitimas_feridas)
rules_vitimas_feridas = getAssociationRules(frequent_itemsets)
rules_vitimas_feridas = rules_vitimas_feridas.sort_values(by='lift', ascending=False)
rules_vitimas_feridas[['antecedents', 'consequents', 'confidence', 'lift', 'leverage', 'conviction']].query('confidence>=0.60').reset_index(drop=True)

antecedents       consequents  confidence      lift  leverage  \
0   (ensolarado, dia útil)             (dia)    0.660830  1.074899  0.025181   
1                (colisão)  (dia útil, reta)    0.609595  1.074810  0.024682   
2         (dia útil, reta)         (colisão)    0.625230  1.074810  0.024682   
3   (ensolarado, dia útil)         (colisão)    0.617033  1.060720  0.019316   
4      (dia útil, colisão)            (reta)    0.720159  1.059761  0.019997   
5                (colisão)            (reta)    0.719559  1.058877  0.023274   
6                   (reta)         (colisão)    0.615961  1.058877  0.023274   
7         (dia útil, reta)      (ensolarado)    0.689973  1.053301  0.019803   
8   (ensolarado, dia útil)            (reta)    0.715585  1.053030  0.019707   
9      (dia útil, colisão)             (dia)    0.646206  1.051111  0.015472   
10         (dia, dia útil)      (ensolarado)    0.688203  1.050598  0.017405   
11                  (reta)      (ensolarado)    0.687916  1.050160  0.022328   
12            (ensolarado)            (reta)    0.713635  1.050160  0.022328   
13          (dia, colisão)        (dia útil)    0.872916  1.048796  0.014804   
14            (ensolarado)             (dia)    0.644279  1.047977  0.019321   
15                   (dia)      (ensolarado)    0.686486  1.047977  0.019321   
16     (dia útil, colisão)      (ensolarado)    0.685280  1.046137  0.014881   
17         (dia, dia útil)         (colisão)    0.605952  1.041670  0.012729   
18               (colisão)      (ensolarado)    0.682275  1.041548  0.015832   
19            (ensolarado)         (colisão)    0.605880  1.041548  0.015832   
20             (dia, reta)        (dia útil)    0.860005  1.033283  0.011325   
21       (ensolarado, dia)        (dia útil)    0.856281  1.028808  0.010119   
22                   (dia)        (dia útil)    0.854144  1.026241  0.013427   
23              (dia útil)             (dia)    0.630916  1.026241  0.013427   
24   (ensolarado, colisão)        (dia útil)    0.850201  1.021504  0.007103   
25               (colisão)             (dia)    0.626630  1.019269  0.006891   
26         (colisão, reta)        (dia útil)    0.847178  1.017872  0.006226   
27               (colisão)        (dia útil)    0.846472  1.017023  0.008242   
28        (dia útil, reta)             (dia)    0.619894  1.008312  0.002898   
29      (ensolarado, reta)        (dia útil)    0.837116  1.005782  0.002250   
30              (dia útil)      (ensolarado)    0.657050  1.003041  0.001658   
31            (ensolarado)        (dia útil)    0.834834  1.003041  0.001658   
32                  (reta)        (dia útil)    0.834620  1.002783  0.001574   
33              (dia útil)            (reta)    0.681440  1.002783  0.001574   
34         (pequeno porte)             (dia)    0.615903  1.001820  0.000560   

    conviction  
0     1.135762  
1     1.108682  
2     1.116119  
3     1.092231  
4     1.145120  
5     1.142668  
6     1.089183  
7     1.112620  
8     1.126704  
9     1.088815  
10    1.106303  
11    1.105284  
12    1.119029  
13    1.319576  
14    1.082917  
15    1.100243  
16    1.096029  
17    1.061515  
18    1.085660  
19    1.061324  
20    1.197876  
21    1.166833  
22    1.149740  
23    1.043710  
24    1.119478  
25    1.031727  
26    1.097335  
27    1.092287  
28    1.013444  
29    1.029547  
30    1.005808  
31    1.015323  
32    1.014006  
33    1.005937  
34    1.002913

> <p style='text-align: justify;'><font size=2.85>Para o cluster de vítimas feridas, temos a principal regra com um lift de 1.06 indicando que quando ocorrem colisões em pistas retas, normalmente durante a semana, é provável que os envolvidos sofram algum tipo de ferimento. Também temos nesse grupo que a maioria dos acidentes que causam ferimentos ocorrem em condições de dias ensolarados ou durante o período do dia. Acidentes desse cluster envolvem veículos de pequeno porte.</font></p>

### Acidentes com vítimas fatais

In [21]:
vitimas_fatais = transactionEncoderData(vitimas_fatais.values)
frequent_itemsets = getFrequentItemsets(vitimas_fatais)
rules_vitimas_fatais = getAssociationRules(frequent_itemsets)
rules_vitimas_fatais = rules_vitimas_fatais.sort_values(by='lift', ascending=False)
rules_vitimas_fatais[['antecedents', 'consequents', 'confidence', 'lift', 'leverage', 'conviction']].query('confidence>=0.60').reset_index(drop=True)

antecedents   consequents  confidence      lift  leverage  \
0                  (noite)        (reta)    0.735494  1.054720  0.019965   
1                    (dia)     (colisão)    0.641352  1.052148  0.015156   
2                    (dia)  (ensolarado)    0.696116  1.044407  0.014112   
3   (ensolarado, dia útil)        (reta)    0.725558  1.040472  0.014785   
4                   (reta)  (ensolarado)    0.692548  1.039053  0.018151   
5             (ensolarado)        (reta)    0.724568  1.039053  0.018151   
6         (dia útil, reta)  (ensolarado)    0.690711  1.036297  0.013313   
7                    (dia)    (dia útil)    0.815701  1.031698  0.011949   
8          (pequeno porte)        (reta)    0.710476  1.018843  0.005890   
9                (colisão)        (reta)    0.706637  1.013339  0.005670   
10                  (reta)     (colisão)    0.617695  1.013339  0.005670   
11        (dia útil, reta)     (colisão)    0.617491  1.013004  0.004362   
12     (dia útil, colisão)        (reta)    0.704637  1.010470  0.003521   
13               (colisão)    (dia útil)    0.791120  1.000608  0.000293   
14              (dia útil)     (colisão)    0.609935  1.000608  0.000293   

    conviction  
0     1.144262  
1     1.088631  
2     1.097398  
3     1.102835  
4     1.084661  
5     1.098873  
6     1.078221  
7     1.135985  
8     1.045385  
9     1.031707  
10    1.021268  
11    1.020724  
12    1.024719  
13    1.002301  
14    1.000950

> <p style='text-align: justify;'><font size=2.85>Em acidentes com vítimas fatais, a nossa principal regra, com lift de 1.05, nos mostra que nesse cluster os acidentes ocorrem, principalmente, durante a noite e em pistas com traçado reto. Boa parte também ocorrem durante a semana, no período do dia e em dias ensolarados. O tipo de acidente mais recorrente são as colisões e essas ocorrências também envolvem veículos de pequeno porte.</font></p>